base/reduce.jl

```diff
@@ -362,10 +367,95 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
+end
+
+function reduce(op, itr, et, isize; kw...)
+    # Fallback: if nothing interesting is being done with the traits
+    # or the operation
+    return mapreduce(identity, op, itr; kw...)
+end
 
 reduce(op, a::Number) = a  # Do we want this?
 
+##### Operation specific reduce optimisations
+
+## vcat
+
+function reduce(::typeof(vcat), xs, T::Type{<:AbstractVector}, isize)
+    x_state = iterate(xs)
+    x_state === nothing && return T() # New empty instance
+    x1, state = x_state
+    
+    ret = copy(x1)  # this is **Not** going to work for StaticArrays
+    
+    if !(isize isa SizeUnknown)
+        # Assume first element has representitive size, unless that would make this too large
+        SIZEHINT_CAP = 10^6
+        sizehint!(ret, min(SIZEHINT_CAP, length(xs)*length(x1)))
```

Comment:
This is a substatial point in speeding this up.
It is the reason knowing the size gives a speedup for vcat.

And it think it is a very common case that 
the size of your vectors are all the same,
or when it is wrong that it is still a good 

I am not sure what a good value for `SIZEHINT_CAP` is.
We need  it to catch cases that only fit in memory because  
the first element is massive, and the rest much  smaller.

Reply:
Are you sure it makes a large difference? IIRC `append!` is supposed to double the size of the storage to ensure resizing doesn't happen too often. At any rate, I don't think it's correct to assume the first element is representative.

Reply:
Yes, especially on small cases it really matters, adding it got a 60% speedup.
one my testcast of 100 vectors of length 128.

Doubling just doesn't increase the capacity that much, if you are doubling a realitvely small number.
Consider if you have 100 vectors of equal length.
Then doubling results in, having to allocate 7 times.
Which is a lot, as a portion of the time you need to spend

So the case of thinking the first element is representative is
is a bit of a guess.
 - If the guess is approximately right then you'll probably only end up doing 1 more allocation.
 - If the guess is too low, then your basically back into the case of doubling, so you've lost nothing by taking it.
 - If the guess is too large, this is the dangerous point, because it risks allocating a ton of memory that isn't needed. 

The last case is where the `SIZEHINT_CAP` comes in. It is our guard against that case. So we set it to some suitable number, I thought `10^6` might be OK, so that would be allocating 4MB if it was `Float64`s. But we could do `10^5` if we anted to be more conservative.
The other thing is that once things are bigger than `SIZEHINT_CAP`, 
then the array size should be large enough that doubling is a very effective increase.

Reply:
I don't think allocating this much extra is reasonable so you need to `sizehint!` back to the actual number in the end (which will reduce the capacity).

Also, I don't think the allocation is what is expensive but the call into the runtime (which can be fixed, https://github.com/JuliaLang/julia/issues/24909). You could check this by calling `resize!` before calling the function.

Also, posting the benchmarks is a good idea.

Reply:
Benchmark:
With code that bypasses normal array reduce:

 - `data = [rand(max(ceil(Int,96+32(randn())), 0)) for ii in 1:1000]` 
 - Timing with `@btime`, different computer from before. (probably a worse computer for benchmarking as it is a shared server, but still point holds)
 - Multiple rounds  of testing to roughly catch on chance sine we  are using arrays
 - round 1:
      - with hint: 498.927 μs (2 allocations: 782.19 KiB),
      - with out :  728.514 μs (11 allocations: 1.80 MiB) 
 - round2:   
      - with hint: 757.129 μs (4 allocations: 2.08 MiB)
      - with out : 900.870 μs (13 allocations: 2.56 MiB)
 - round3:
      - with hint: 643.748 μs (3 allocations: 1.74 MiB)
      - with out : 687.555 μs (11 allocations: 1.51 MiB)

Now of-course the advantage goes up if your inital guess at the size was wrong.
And once `sizehint!`ing back down is added, that will also cut down the advantage in doing it.
But I think it will still be worth it 
(can further add heuristics about not sizehinting down if <2x too large,  since that is acceptable)

Reply:
@KristofferC  @nalimilan  do the above benchmarks satisfy you?